streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredVersionedKeyValueStore.java

```diff
@@ -0,0 +1,226 @@
+     * {@link VersionedKeyValueStore} interface itself.
+     */
+    private class MeteredVersionedKeyValueStoreInternal
+        extends MeteredKeyValueStore<K, ValueAndTimestamp<V>>
+        implements TimestampedKeyValueStore<K, V> {
+
+        private final VersionedBytesStore inner;
+
+        MeteredVersionedKeyValueStoreInternal(final VersionedBytesStore inner,
+                                              final String metricScope,
+                                              final Time time,
+                                              final Serde<K> keySerde,
+                                              final Serde<ValueAndTimestamp<V>> valueSerde) {
+            super(inner, metricScope, time, keySerde, valueSerde);
+            this.inner = inner;
+        }
+
+        @Override
+        public void put(final K key, final ValueAndTimestamp<V> value) {
+            super.put(
+                key,
+                // versioned stores require a timestamp associated with all puts, including tombstones/deletes
+                value == null
+                    ? ValueAndTimestamp.makeAllowNullable(null, context.timestamp())
+                    : value
+            );
+        }
+
+        public ValueAndTimestamp<V> get(final K key, final long asOfTimestamp) {
+            Objects.requireNonNull(key, "key cannot be null");
+            try {
+                return maybeMeasureLatency(() -> outerValue(inner.get(keyBytes(key), asOfTimestamp)), time, getSensor);
+            } catch (final ProcessorStateException e) {
+                final String message = String.format(e.getMessage(), key);
+                throw new ProcessorStateException(message, e);
+            }
+        }
+
+        public ValueAndTimestamp<V> delete(final K key, final long timestamp) {
+            Objects.requireNonNull(key, "key cannot be null");
+            try {
+                return maybeMeasureLatency(() -> outerValue(inner.delete(keyBytes(key), timestamp)), time, deleteSensor);
+            } catch (final ProcessorStateException e) {
+                final String message = String.format(e.getMessage(), key);
+                throw new ProcessorStateException(message, e);
+            }
+        }
+
+        @Override
+        public <R> QueryResult<R> query(final Query<R> query,
+                                        final PositionBound positionBound,
+                                        final QueryConfig config) {
+            final long start = time.nanoseconds();
+            final QueryResult<R> result = wrapped().query(query, positionBound, config);
+            if (config.isCollectExecutionInfo()) {
+                result.addExecutionInfo(
+                    "Handled in " + getClass() + " in " + (time.nanoseconds() - start) + "ns");
+            }
+            // do not convert return type from inner bytes store to user-friendly interface at this
```

Comment:
This design choice has implications -- I'm curious to hear reviewer thoughts.

(The comment should more accurately say "do not convert query or return types to/from inner bytes store ...", which I will update in the commit itself.)

The existing MeteredKeyValueStore and MeteredTimestampedKeyValueStore both contain logic to assist in translating `KeyQuery<K, V>` and `RangeQuery<K, V>` types to queries understandable by the underlying bytes store, and to deserialize the returned bytes result back for the original `K` and `V` result types. The return type of these queries for MeteredKeyValueStore and MeteredTimestampedKeyValueStore is `V` for the key query, and `KeyValueIterator<K, V>` for the range query. We'll want to introduce new return types for versioned stores, perhaps `VersionedRecord<V>` for the key query and `Iterator<VersionedRecord<V>>` for the range query. I'd like to postpone this decision-making for a follow-up KIP, but that means not assisting in translating to/from the inner bytes store for now.

So we have three options in the meantime:
1. Reject all IQv2 queries, since RocksDBVersionedKeyValueStore doesn't support IQv2 yet anyway. I don't think we should do this, because then users who create their own VersionedKeyValueStore implementations won't be able to issue IQv2 queries either.
2. Allow IQv2 queries as a pass-through at the metered layer. This is what the PR currently does. It means that users who implement their own VersionedKeyValueStores will be able to issue IQv2 queries, but also has backwards compatibility implications for when we introduce type serialization/deserialization support for key and range queries in the future. The behavior at this metered store layer would suddenly change for those query types.
3. Reject KeyQuery and RangeQuery types in order to reserve them for future behaviors, and pass-through all other query types. This avoids compatibility concerns for the future, while still allowing users to issue IQv2 queries to their own VersionedKeyValueStore implementations in the meantime.

I'm curious to know whether there's existing precedent / discussion on this compatibility concern for introducing custom logic at the metered layer for handling specific types of IQv2 queries.